data/src/main/java/com/microsoft/azure/kusto/data/auth/CloudInfo.java

```diff
@@ -0,0 +1,133 @@
+    private final String kustoServiceResourceId;
+    private final String firstPartyAuthorityUrl;
+
+    public CloudInfo(boolean loginMfaRequired, String loginEndpoint, String kustoClientAppId, String kustoClientRedirectUri, String kustoServiceResourceId, String firstPartyAuthorityUrl) {
+        this.loginMfaRequired = loginMfaRequired;
+        this.loginEndpoint = loginEndpoint;
+        this.kustoClientAppId = kustoClientAppId;
+        this.kustoClientRedirectUri = kustoClientRedirectUri;
+        this.kustoServiceResourceId = kustoServiceResourceId;
+        this.firstPartyAuthorityUrl = firstPartyAuthorityUrl;
+    }
+
+    public static void manuallyAddToCache(String clusterUrl, CloudInfo cloudInfo){
+        synchronized (cache) {
+            cache.put(clusterUrl, cloudInfo);
+        }
+    }
+
+    public static CloudInfo retrieveCloudInfoForCluster(String clusterUrl) throws DataServiceException {
+        synchronized (cache) {
+            CloudInfo cloudInfo = cache.get(clusterUrl);
+            if (cloudInfo != null) {
+                return cloudInfo;
+            }
+
+            CloudInfo result;
+
+            try (CloseableHttpClient httpClient = HttpClients.createDefault()) {
+                HttpGet request = new HttpGet(new URIBuilder(clusterUrl).setPath(METADATA_ENDPOINT).build().toString());
+                request.addHeader(HttpHeaders.ACCEPT_ENCODING, "gzip,deflate");
+                request.addHeader(HttpHeaders.ACCEPT, "application/json");
+                try (CloseableHttpResponse response = httpClient.execute(request)) {
+                    int statusCode = response.getStatusLine().getStatusCode();
+                    if (statusCode == 200) {
+                        String content = EntityUtils.toString(response.getEntity());
+                        if (content == null || content.equals("") || content.equals("{}")) {
+                            throw new DataServiceException(clusterUrl, "Error in metadata endpoint, received no data", true);
+                        }
+                        result = parseCloudInfo(content);
+                    } else if (statusCode == 404) {
+                        result = DEFAULT_CLOUD;
+                    } else {
+                        String errorFromResponse = EntityUtils.toString(response.getEntity());
+                        throw new DataServiceException(clusterUrl,
+                                "Error in metadata endpoint, got code: " + statusCode + "\nWith error: " + errorFromResponse, true);
+                    }
+                }
+            } catch (IOException | URISyntaxException ex) {
+                throw new DataServiceException(clusterUrl, "IOError when trying to retrieve CloudInfo", ex, true);
+            }
+            cache.put(clusterUrl, result);
+            return result;
+        }
+    }
+
+    private static CloudInfo parseCloudInfo(String content) {
+        JSONObject jsonObject = new JSONObject(content);
```

Comment:
Ran this with local Engine and got result
{"dSTS":null,"AzureAD":null}
which fails on getJSONObject("AzureAD)
Might want to check this on other SDKs too (return here null and use default)

Reply:
Good catch, also fixed in python